File: src/parallel_execution/parallel_execution.erl
-module(parallel_execution).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/parallel_execution.md"}.

-export([
    run/1
]).

-spec run({unicode:chardata(), unicode:chardata(), unicode:chardata(), unicode:chardata()}) ->
    temporal_sdk:workflow_result() | no_return().
run({_, _, _, _} = Input) ->
    {ok, _StartWorkflowRet, WorkflowResult} =
        temporal_sdk:start_workflow(cluster_1, "default", parallel_execution_workflow, [
            await, {input, [tuple_to_list(Input)]}
        ]),
    WorkflowResult.

File: lib/parallel_execution/parallel_execution.ex
defmodule ParallelExecution do
  @external_resource "docs/parallel_execution.md"
  @moduledoc TemporalSdk.Utils.Code.exdoc!("docs/parallel_execution.md")

  @spec run({String.t(), String.t(), String.t(), String.t()}) ::
          :temporal_sdk.workflow_result() | no_return()
  def run({_, _, _, _} = input) do
    {:ok, _start_workflow_ret, workflow_result} =
      TemporalSdk.start_workflow(:cluster_1, "default", ParallelExecution.Workflow, [
        :await,
        input: [Tuple.to_list(input)]
      ])

    workflow_result
  end
end
